website/server/controllers/api-v3/members.js

```diff
@@ -402,6 +362,109 @@ function _getMembersForItem (type) {
+  });
+  if (!group || !challenge.canView(user, group)) throw new NotFound(res.t('challengeNotFound'));
+  if (!challenge.isMember(member)) throw new NotFound(res.t('challengeMemberNotFound'));
+
+  const challengeTasks = await Tasks.Task.find({
+    userId: memberId,
+    'challenge.id': challengeId,
+  })
+    .select('-tags') // We don't want to return the tags publicly TODO same for other data?
+    .exec();
+
+  return challengeTasks.map(task => {
+    task.checklist = []; // Clear checklists as they are private
+    return task.toJSON({ minimize: true });
+  });
+}
+
+async function _handleGetMembersForChallenge (req, res) {
+  req.checkParams('challengeId', res.t('challengeIdRequired')).notEmpty().isUUID();
+  req.checkQuery('lastId').optional().notEmpty().isUUID();
+  // Allow an arbitrary number of results (up to 60)
+  req.checkQuery('limit', res.t('groupIdRequired')).optional().notEmpty().isInt({ min: 1, max: 60 });
+
+  const validationErrors = req.validationErrors();
+  if (validationErrors) throw validationErrors;
+
+  const { challengeId } = req.params;
+  const { lastId } = req.query;
+  const { user } = res.locals;
+
+  const challenge = await Challenge.findById(challengeId).select('_id type leader group').exec();
+  if (!challenge) throw new NotFound(res.t('challengeNotFound'));
+
+  // optionalMembership is set to true because even
+  // if you're not member of the group you may be able to access the challenge
+  // for example if you've been booted from it, are the leader or a site admin
+  const group = await Group.getGroup({
+    user,
+    groupId: challenge.group,
+    fields: '_id type privacy',
+    optionalMembership: true,
+  });
+
+  if (!group || !challenge.canView(user, group)) throw new NotFound(res.t('challengeNotFound'));
+
+  const query = {};
+  let fields = nameFields;
+  // add computes stats to the member info when items and stats are available
+  let addComputedStats = false;
+  query.challenges = challenge._id;
+
+  if (req.query.includeAllPublicFields === 'true') {
+    fields = memberFields;
+    addComputedStats = true;
+  }
+
+  if (req.query.search) {
+    query['auth.local.username'] = { $regex: req.query.search };
+  }
+
+  if (lastId) query._id = { $gt: lastId };
+
+  let limit = req.query.limit ? Number(req.query.limit) : 30;
+
+  // Allow for all challenges members to be returned
+  if (req.query.includeAllMembers === 'true') {
+    limit = 0; // no limit
+  }
+
+  const members = await User
+    .find(query)
+    .sort({ _id: 1 })
+    .limit(limit)
+    .select(fields)
+    .lean()
+    .exec();
+
+  // manually call toJSON with minimize: true so empty paths aren't returned
+  members.forEach(member => User.transformJSONUser(member, addComputedStats));
+  if (req.query.includeTasks === 'true') {
+    const tasksPremises = members.map(
+      member => _getMemberTasksFromChallenge(challengeId, member.id, user, res),
```

Comment:
I'd like to change this so that instead of calling `_getMemberTasksFromChallenge ` for every challenge member (creating a lot of promises and database queries) if we could transform it so that only one call is necessary, as I said challenge and group data has already been fetched, same for members.

For the tasks it'd be enough to make a single query with `userId:{$in: anArrayOfAllMembersIds}, `

Reply:
Just to clarify, each member of a challenge has the same set of tasks related to the challenge, so there is no need to go over each member and fetch his tasks separately, right?